docs/html/reference/requirements-file-format.md

```diff
@@ -56,9 +56,15 @@ examples of all these forms, see {ref}`pip install Examples`.
 
 ### Encoding
 
-Requirements files are `utf-8` encoding by default and also support
-{pep}`263` style comments to change the encoding (i.e.
-`# -*- coding: <encoding name> -*-`).
+It is simplest to encode your requirements files with UTF-8.
+The process for decoding requirements files is:
+
+- Check for any Byte Order Mark at the start of the file and if found use
+  the corresponding encoding to decode the file.
+- Check for any {pep}`263` style comment (e.g. `# -*- coding: <encoding name> -*-`)
+  and if found decode with the given encoding.
+- Try and decode with UTF-8, and if that fails,
+- fallback to trying to decode using the locale defined encoding.
```

Comment:
This is the specification for requirement files, we should be tighter. 
````md
The default encoding for requirement files is `UTF-8` unless a different encoding
is specified using a {pep}`263` style comment (e.g. `# -*- coding: <encoding name> -*-`). 

```{warning}
pip will fallback to the locale defined encoding if `UTF-8` decoding fails. This is a quirk
of pip's parser. This behaviour is *deprecated* and should not be relied upon.
```
````

While we're here, should we expand the specification to formally allow BOMs?

Reply:
> While we're here, should we expand the specification to formally allow BOMs?

How much do we want to encourage this usage, do you have an idea for how much they are relied on? Specifically, I'm wondering if we can just say: UTF-8 by default, PEP-263 style comments supported, other bits deprecated (if someone is relying on locale specific or BOM, they should switch to PEP-263 comment)

Reply:
I would prefer to *not* allow BOMs, as @matthewhughes934 says. Ideally, as it's a backward incompatible change, we would initially issue a deprecation warning if a BOM is found, but given that the use of a BOM is undocumented (in the [user docs](https://pip.pypa.io/en/stable/reference/requirements-file-format/#encoding)), I'd be fine with simply removing that support if that's easier.

However, PEP 263 states that the UTF8 signature `'\xef\xbb\xbf'` should signal UTF8, so that one (which is treated as a BOM in `auto_decode`) should stay.

It's worth noting that UTF16 and UTF32 encodings can't be handled with a PEP 263 style comment, as those encodings are not ASCII supersets. But I think it's fine to stop supporting them.

Reply:
> It's worth noting that UTF16 and UTF32 encodings can't be handled with a PEP 263 style comment, as those encodings are not ASCII supersets. But I think it's fine to stop supporting them.

I'm not following this issue closely, but in case it's helpful, I'll note that uv initially only supported UTF8 but then found they had user issues not supporting UTF16: https://github.com/astral-sh/uv/pull/2283

I think because redirecting from stdout to a file on Windows would end up with a UTF16 file, e.g. `pip freeze > requirements.txt`.

Reply:
I'm not really in the mood to break our Windows users over requirements file decoding, honestly. I'd prefer documenting BOMs as supported as I don't like to have implementation-defined behaviour here, but I also recognize that would _promote_ their proliferation so leaving this as implementation-defined behaviour is a fair compromise. Using the old docs text or my suggestion (without any changes) is fine. 

Thanks @notatallshaw for letting us know that UTF-16 can be pretty common on Windows.

Reply:
> I'm not really in the mood to break our Windows users over requirements file decoding, honestly. I'd prefer documenting BOMs as supported as I don't like to have implementation-defined behaviour here, but I also recognize that would _promote_ their proliferation so leaving this as implementation-defined behaviour is a fair compromise. Using the old docs text or my suggestion (without any changes) is fine.

:+1: I like your more concise docs, and the warning, so went with that 5c49381e4

Reply:
> I think because redirecting from stdout to a file on Windows would end up with a UTF16 file

In Windows 11, cmd seems to use UTF8[^1], as does Powershell Core. Windows Powershell does use UTF16, though, so I guess there's a proportion of users who will encounter this (I don't have a feel for the relative usage of the 3 shells).

Given this, +1 on retaining the existing support but not documenting it.

[^1]: This may depend on the console codepage, I'm not sure when UTF8 became the default.

Reply:
:+1: I added a code comment explaining why we still support BOMs: https://github.com/pypa/pip/pull/12795/commits/f1dbf496677b950680ceed74a76cccaf0fe6cf17